bpfd-operator/controllers/bpfd-agent/uprobe-program.go

```diff
@@ -0,0 +1,340 @@
+	// the same kernel ProgramType, we use internal.Kprobe below.
+	programMap, err := bpfdagentinternal.ListBpfdPrograms(ctx, r.BpfdClient, internal.Kprobe)
+	if err != nil {
+		r.Logger.Error(err, "failed to list loaded bpfd programs")
+		return ctrl.Result{Requeue: true, RequeueAfter: retryDurationAgent}, nil
+	}
+
+	// Reconcile each UprobeProgram. Don't return error here because it will trigger an infinite reconcile loop, instead
+	// report the error to user and retry if specified. For some errors the controller may not decide to retry.
+	// Note: This only results in grpc calls to bpfd if we need to change something
+	requeue := false // initialize requeue to false
+	for _, uprobeProgram := range uprobePrograms.Items {
+		r.Logger.Info("UprobeProgramController is reconciling", "currentUprobeProgram", uprobeProgram.Name)
+		r.currentUprobeProgram = &uprobeProgram
+		result, err := reconcileProgram(ctx, r, r.currentUprobeProgram, &r.currentUprobeProgram.Spec.BpfProgramCommon, r.ourNode, programMap)
+		if err != nil {
+			r.Logger.Error(err, "Reconciling UprobeProgram Failed", "UprobeProgramName", r.currentUprobeProgram.Name, "ReconcileResult", result.String())
+		}
+
+		switch result {
+		case internal.Unchanged:
+			// continue with next program
+		case internal.Updated:
+			// return
+			return ctrl.Result{Requeue: false}, nil
+		case internal.Requeue:
+			// remember to do a requeue when we're done and continue with next program
+			requeue = true
+		}
+	}
+
+	if requeue {
+		// A requeue has been requested
+		return ctrl.Result{RequeueAfter: retryDurationAgent}, nil
+	} else {
+		// We've made it through all the programs in the list without anything being
+		// updated and a reque has not been requested.
+		return ctrl.Result{Requeue: false}, nil
+	}
+}
+
+func (r *UprobeProgramReconciler) buildUprobeLoadRequest(
+	bytecode interface{},
+	uuid string,
+	bpfProgram *bpfdiov1alpha1.BpfProgram,
+	mapOwnerId *uint32) *gobpfd.LoadRequest {
+	loadRequest := &gobpfd.LoadRequest{}
+	loadRequest.Common = bpfdagentinternal.BuildBpfdCommon(
+		bytecode,
+		r.currentUprobeProgram.Spec.SectionName,
+		internal.Kprobe, map[string]string{internal.UuidMetadataKey: uuid},
+		r.currentUprobeProgram.Spec.GlobalData,
+		mapOwnerId,
+	)
+	// Namespace isn't supported yet in bpfd, so set it to an empty string.
+	namespace := ""
+	loadRequest.AttachInfo = &gobpfd.LoadRequest_UprobeAttachInfo{
+		UprobeAttachInfo: &gobpfd.UprobeAttachInfo{
+			FnName:    &r.currentUprobeProgram.Spec.FunctionName,
+			Offset:    r.currentUprobeProgram.Spec.Offset,
+			Target:    bpfProgram.Annotations[internal.UprobeProgramTarget],
+			Retprobe:  r.currentUprobeProgram.Spec.RetProbe,
+			Namespace: &namespace,
+		},
+	}
+
+	return loadRequest
+}
+
+// reconcileBpfdPrograms ONLY reconciles the bpfd state for a single BpfProgram.
+// It does not interact with the k8s API in any way.
+func (r *UprobeProgramReconciler) reconcileBpfdProgram(ctx context.Context,
```

Comment:
You don't have to deal with this now... but it __should__ be farily simple to move all of this code into common.go the more I look at it.. That's a bit smaller diff everytime we add a new program :celebrate

Reply:
Agreed.  I opened https://github.com/bpfd-dev/bpfd/issues/709.